pygmt/helpers/utils.py

```diff
@@ -550,3 +551,37 @@ def args_in_kwargs(args, kwargs):
     return any(
         kwargs.get(arg) is not None and kwargs.get(arg) is not False for arg in args
     )
+
+
+def return_table(session, output_type, vfile, colnames):
+    """
+    Return an output table from a virtual file based on the output type.
+
+    Parameters
+    ----------
+    session : :class:`pygmt.clib.Session`
+        The current session.
+    output_type : str
+        The output type. Can be ``"pandas"``, ``"numpy"``, or ``"file"``.
+    vfile : str
+        The virtual file name.
+    colnames : list of str
+        The column names for the :class:`pandas.DataFrame` output.
+
+    Returns
+    -------
+    :class:`pandas.DataFrame` or :class:`numpy.ndarray` or None
+        The output table. If ``output_type`` is ``"file"``, returns ``None``.
+    """
+    if output_type == "file":  # Already written to file, so return None
+        return None
+    # Read the virtual file as a GMT dataset and convert to vectors
+    vectors = session.read_virtualfile(vfile, kind="dataset").contents.to_vectors()
+    # pandas.DataFrame output
+    if colnames is None:
+        colnames = pd.RangeIndex(0, len(vectors))
+    result = pd.DataFrame.from_dict(dict(zip(colnames, vectors)))
```

Comment:
`vectors` is a list of vectors and each vector represents a column in the data. I've tried different solutions to create a pd.DataFrame object, but most are not ideal.

Here is an example:
```
>>> import numpy as np
>>> import pandas as pd
>>> vectors = [np.array([1, 2, 3]), [4.0, 5.0, 6.0], [7.0, 8.0, 9.0], ["AA", "BB", "CC"]]
>>> colnames = ["col1", "col2", "col3", "col4"]
```

```
# Each vector is taken as a row, not a column
>>> df = pd.DataFrame(vectors)
>>> df
     0    1    2
0    1    2    3
1  4.0  5.0  6.0
2  7.0  8.0  9.0
3   AA   BB   CC

# The dtype information is lost 
>>> df = pd.DataFrame(np.array(vectors).T)
>>> df
   0    1    2   3
0  1  4.0  7.0  AA
1  2  5.0  8.0  BB
2  3  6.0  9.0  CC
>>> df.dtypes
0    object
1    object
2    object
3    object
dtype: object

# It's weird to set column names via `index` and again dtype is lost
>>> df = pd.DataFrame(vectors, index=colnames).T
>>> df
  col1 col2 col3 col4
0    1  4.0  7.0   AA
1    2  5.0  8.0   BB
2    3  6.0  9.0   CC
>>> df.dtypes
col1    object
col2    object
col3    object
col4    object
dtype: object

# Current best solution
>>> df = pd.DataFrame.from_dict(dict(zip(colnames, vectors)))
>>> df
   col1  col2  col3 col4
0     1   4.0   7.0   AA
1     2   5.0   8.0   BB
2     3   6.0   9.0   CC
>>> df.dtypes
col1      int64
col2    float64
col3    float64
col4     object
dtype: object
```

Reply:
> If we want to turn that string column's object dtype to a proper `string` dtype, we can use pandas's [`convert_dtypes`](https://pandas.pydata.org/pandas-docs/version/2.1/reference/api/pandas.DataFrame.convert_dtypes.html):

It's likely `convert_dtypes` will cause more trouble for us:
```python
>>> df = pd.DataFrame.from_dict(dict(zip(colnames, vectors)))
>>> df["col1"].to_numpy()
array([1, 2, 3])
>>> df["col1"].to_numpy().dtype
dtype('int64')

>>> df = df.convert_dtypes(dtype_backend="numpy_nullable")
>>> df["col1"].to_numpy()
array([1, 2, 3], dtype=object)
```
As you can see, the `to_numpy` function now returns a `object` dtype instead of a `int64` type.

Reply:
We could also just convert string dtypes, and avoid converting integer and float into nullable dtypes like so:

```python
df = pd.DataFrame.from_dict(dict(zip(colnames, vectors)))
df = df.convert_dtypes(
    convert_integer=False,
    convert_floating=False,
    convert_string=True,
    dtype_backend="numpy_nullable",
)
df.dtypes
# col1             int64
# col2           float64
# col3           float64
# col4    string[python]
# dtype: object
df["col1"].to_numpy().dtype
# dtype('int64')
```

Reply:
Done in b9454ce0c29e83b194650906704fa1aa26f398f8. 

[`convert_dtypes`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.convert_dtypes.html)'s `convert_floating` parameter is only for pandas >= 1.2.0. Do we need to add `pandas>=1.2.0` in `pyproject.toml`?

Reply:
[Pandas 1.2.0 was released on 26 Dec 2020](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.2.0.html), or about 3 years ago. I tried to create a conda environment with pandas 1.1.x

```
mamba create -n temp 'pandas~=1.1.0' 'numpy==1.22' 'python==3.9'
```

and it seemed to be possible

```
Looking for: ['pandas~=1.1.0', 'numpy==1.23', 'python==3.9']

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache
Transaction

  Prefix: /home/user/mambaforge/envs/temp

  Updating specs:

   - pandas ~=1.1.0
   - numpy==1.23
   - python==3.9


  Package                  Version  Build                Channel                    Size
──────────────────────────────────────────────────────────────────────────────────────────
  Install:
──────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex              0.1  conda_forge          conda-forge/linux-64     Cached
  + _openmp_mutex              4.5  2_gnu                conda-forge/linux-64     Cached
  + ca-certificates      2023.7.22  hbcca054_0           conda-forge/linux-64     Cached
  + ld_impl_linux-64          2.40  h41732ed_0           conda-forge/linux-64     Cached
  + libblas                  3.9.0  19_linux64_openblas  conda-forge/linux-64     Cached
  + libcblas                 3.9.0  19_linux64_openblas  conda-forge/linux-64     Cached
  + libffi                     3.3  h58526e2_2           conda-forge/linux-64       53kB
  + libgcc-ng               13.2.0  h807b86a_3           conda-forge/linux-64     Cached
  + libgfortran-ng          13.2.0  h69a702a_3           conda-forge/linux-64     Cached
  + libgfortran5            13.2.0  ha4646dd_3           conda-forge/linux-64     Cached
  + libgomp                 13.2.0  h807b86a_3           conda-forge/linux-64     Cached
  + liblapack                3.9.0  19_linux64_openblas  conda-forge/linux-64     Cached
  + libopenblas             0.3.24  pthreads_h413a1c8_0  conda-forge/linux-64     Cached
  + libsqlite               3.44.0  h2797004_0           conda-forge/linux-64     Cached
  + libstdcxx-ng            13.2.0  h7e041cc_3           conda-forge/linux-64     Cached
  + libzlib                 1.2.13  hd590300_5           conda-forge/linux-64     Cached
  + ncurses                    6.4  h59595ed_2           conda-forge/linux-64     Cached
  + numpy                   1.23.0  py39hba7629e_0       conda-forge/linux-64        7MB
  + openssl                 1.1.1w  hd590300_0           conda-forge/linux-64        2MB
  + pandas                   1.1.5  py39hde0f152_0       conda-forge/linux-64       12MB
  + pip                     23.3.1  pyhd8ed1ab_0         conda-forge/noarch       Cached
  + python                   3.9.0  hffdb5ce_5_cpython   conda-forge/linux-64       30MB
  + python-dateutil          2.8.2  pyhd8ed1ab_0         conda-forge/noarch       Cached
  + python_abi                 3.9  4_cp39               conda-forge/linux-64        6kB
  + pytz              2023.3.post1  pyhd8ed1ab_0         conda-forge/noarch       Cached
  + readline                   8.2  h8228510_1           conda-forge/linux-64     Cached
  + setuptools              59.8.0  py39hf3d152e_1       conda-forge/linux-64        1MB
  + six                     1.16.0  pyh6c4a22f_0         conda-forge/noarch       Cached
  + sqlite                  3.44.0  h2c6b66d_0           conda-forge/linux-64     Cached
  + tk                      8.6.13  noxft_h4845f30_101   conda-forge/linux-64     Cached
  + tzdata                   2023c  h71feb2d_0           conda-forge/noarch       Cached
  + wheel                   0.41.3  pyhd8ed1ab_0         conda-forge/noarch       Cached
  + xz                       5.2.6  h166bdaf_0           conda-forge/linux-64     Cached
  + zlib                    1.2.13  hd590300_5           conda-forge/linux-64     Cached

  Summary:

  Install: 34 packages

  Total download: 53MB
```

So yeah, we might need to set a lower pin on pandas>=1.2.0 if using `convert_floating=False`. I don't think there's a way to make this backwards compatible with pandas 1.1.x since `convert_floating` is `True` by default.

Reply:
> we might need to set a lower pin on pandas>=1.2.0 if using `convert_floating=False`.

Done in b7e082347de9a009ef314596a199207681c6b154.